lib/tasks/manageiq/ui_tasks.rake

```diff
@@ -24,6 +24,100 @@ namespace :update do
   task :ui => ['update:clean', 'update:bower', 'update:yarn', 'webpack:compile']
 end
 
+namespace :ui do
+  desc "Clean up node_modules and bower assets for all stale ui_classic gems on the system"
+  task :clean do
+    require 'fileutils'
+
+    git_dirs       = Dir["#{Gem.dir}/bundler/gems/*"].grep(/manageiq-ui-classic/)
+    spec_git_paths = Bundler.definition.spec_git_paths.grep(/manageiq-ui-classic/)
+    stale_ui_gems  = (git_dirs - spec_git_paths)
+    total_savings  = 0
+
+    warn "DRY RUN:  NO OPERATIONS BEING EXECUTED!!" if ENV["DRY_RUN"]
+
+    stale_ui_gems.each do |stale_miq_ui_gem_dir|
+      [
+        File.join(stale_miq_ui_gem_dir, "node_modules"),
+        File.join(stale_miq_ui_gem_dir, "vendor", "assets")
```

Comment:
can this be +`, "bower"`?

We did have some vendor files which were not bower at some point..

Reply:
@himdel Addressed this in my rebase, and I think the `gem_regexp` should do the trick... but the sad part was:

```
DRY RUN:  NO OPERATIONS BEING EXECUTED!!
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-graphql-4b9ec2244e05/node_modules... (15.6MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-graphql-55a4ec981f49/node_modules... (20.6MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-graphql-d08c7ff58785/node_modules... (14.2MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-265ba60b2652/node_modules... (193.8MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-265ba60b2652/vendor/assets/bower... (123.8MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-61026e3296a1/node_modules... (195.3MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-61026e3296a1/vendor/assets/bower... (123.8MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-6805dd8a1aa7/node_modules... (143.7MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-a0ff186c3fef/node_modules... (186.2MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-a0ff186c3fef/vendor/assets/bower... (123.8MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-a63c3559bb12/node_modules... (186.2MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-a63c3559bb12/vendor/assets/bower... (123.8MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/miq_v2v_ui_plugin-761f3ad07ef3/node_modules... (206.5MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/miq_v2v_ui_plugin-cd901ff708d5/node_modules... (204.1MB)

Total Savings: 1.8GB
```

And I cleaned this up like 3 weeks ago... 😭

Reply:
Ouch :)

The fun thing is, it only happens when using the gems, so I mostly blame bundler for putting each version in a new directory and not cleaning up the rest. (That said, it's definitely wrong to install `node_modules` under a gem dir too.)

My personal solution to this has been:

`30 14 * * * hardlink -pot ~/.rbenv/ ~/ManageIQ/`

in my `crontab`, but looks like we should really look into moving all the `node_modules` under `manageiq`.

(My guess about the dependency ordering here will be: finish the move to webpack5, try to do that, realize npm can't actually install to any directory other than <where package.json is>/node_modules and while require/import have the same problem this can be overriden in weback loader configs .. which will likely only work for JS (babel), not typescript, and then spend some time figuring out whether typescript loaders can do the same or whether we need to pass ts through babel as well and just use tsc as a linter :).)